compositor_render/src/renderer/texture.rs

```diff
@@ -32,4 +38,225 @@ impl Texture {
     pub fn size(&self) -> wgpu::Extent3d {
         self.texture.size()
     }
+
+    pub fn new_rgba(ctx: &WgpuCtx, resolution: &Resolution) -> Self {
+        Self::new(
+            ctx,
+            None,
+            wgpu::Extent3d {
+                width: resolution.width as u32,
+                height: resolution.height as u32,
+                depth_or_array_layers: 1,
+            },
+            wgpu::TextureFormat::Rgba8Unorm,
+            wgpu::TextureUsages::RENDER_ATTACHMENT
+                | wgpu::TextureUsages::TEXTURE_BINDING
+                | wgpu::TextureUsages::COPY_DST,
+        )
+    }
+
+    pub fn new_yuv_textures(ctx: &WgpuCtx, resolution: &Resolution) -> [Texture; 3] {
+        let y = Self::new(
+            ctx,
+            None,
+            wgpu::Extent3d {
+                width: resolution.width as u32,
+                height: resolution.height as u32,
+                depth_or_array_layers: 1,
+            },
+            wgpu::TextureFormat::R8Unorm,
+            wgpu::TextureUsages::RENDER_ATTACHMENT
+                | wgpu::TextureUsages::COPY_SRC
+                | wgpu::TextureUsages::COPY_DST,
+        );
+        let u = Self::new(
+            ctx,
+            None,
+            wgpu::Extent3d {
+                width: resolution.width as u32 / 2,
+                height: resolution.height as u32 / 2,
+                depth_or_array_layers: 1,
+            },
+            wgpu::TextureFormat::R8Unorm,
+            wgpu::TextureUsages::RENDER_ATTACHMENT
+                | wgpu::TextureUsages::COPY_SRC
+                | wgpu::TextureUsages::COPY_DST,
+        );
+        let v = Self::new(
+            ctx,
+            None,
+            wgpu::Extent3d {
+                width: resolution.width as u32 / 2,
+                height: resolution.height as u32 / 2,
+                depth_or_array_layers: 1,
+            },
+            wgpu::TextureFormat::R8Unorm,
+            wgpu::TextureUsages::RENDER_ATTACHMENT
+                | wgpu::TextureUsages::COPY_SRC
+                | wgpu::TextureUsages::COPY_DST,
+        );
+        [y, u, v]
+    }
+
+    pub fn upload_frame_to_textures(ctx: &WgpuCtx, textures: &[Texture; 3], frame: Arc<Frame>) {
+        // TODO maybe resize?
+        textures[0].upload_data(&ctx.queue, &frame.data.y_plane, 1);
+        textures[1].upload_data(&ctx.queue, &frame.data.u_plane, 1);
+        textures[2].upload_data(&ctx.queue, &frame.data.v_plane, 1);
+        ctx.queue.submit([]);
```

Comment:
Can we add a note here saying there is an optimization opportunity here? Apparently (according to [this](https://github.com/gfx-rs/wgpu/wiki/Do's-and-Dont's#dont-submit-many-times-per-frame)) submits can be CPU-intensive and I would like to reduce the amount of them after we make an initial version of the compositor. Wouldn't want to miss this then :)

Reply:
thanks for the link, yes I agree this could will still need some work regardless, because we need to support dynamic texture rescaling either way